### PR TITLE
Emit OpMemberDecorate for member 0 annotations

### PR DESCRIPTION
**Pull request by dwoodwor-intel**
LLVMToSPIRVBase::transIntrinsicInst translates ptr.annotation intrinsics marking struct members by looking for struct GEPs. However, a GEPs is not necessary for the first member, and so in some cases when the first member is annotated SPIRVWriter has been emitting OpDecorate for these instead which leads to a crash in some cases on reverse translation. This change enables emitting OpMemberDecorate for the first member even if there isn't a GEP and the pointer value is based on a struct alloca instead.

This change also moves the handling of FPGARegIntel annotations so that they're translated to OpFPGARegINTEL before attempting to translate struct member annotations. This ensures that they don't get incorrectly translated with OpMemberDecorate.

### CODE REVIEW

**Comment by dwoodwor-intel** [comment]
This is another approach to the problem in #1635 which seems to be a more proper fix

File: lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -3745,15 +3745,38 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
...
-      auto *Ty = transType(GI->getSourceElementType());
-      auto *ResPtr = transValue(GI, BB);
-      SPIRVWord MemberNumber =
-          dyn_cast<ConstantInt>(GI->getOperand(2))->getZExtValue();
+    // for the GEP itself. There may not be a GEP in this case if the access is
+    // to the first member of the struct; if so, attempt to get a struct type
+    // from an alloca instead.
+    SPIRVType *StructTy = nullptr;
+    SPIRVValue *ResPtr;
+    SPIRVWord MemberNumber;
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVWriter.cpp:3765]:
```suggestion
    [[maybe_unused]] SPIRVValue *ResPtr = nullptr;
    [[maybe_unused]] SPIRVWord MemberNumber = 0;
```

**Review by MrSidims — COMMENTED** [review]
LGTM just one nit

**Comment by MrSidims** [comment]
@asudarsa @vmaksimo could you please also take a look?

File: lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -3745,15 +3745,38 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
     AnnotationDecorations Decorations =
         tryParseAnnotationString(BM, AnnotationString);
 
+    // Translate FPGARegIntel annotations to OpFPGARegINTEL.
```
Comment (asudarsa) [inline lib/SPIRV/SPIRVWriter.cpp:3748]: This change seems to be unrelated to the issue being addressed. Can you please comment? Thanks

File: lib/SPIRV/SPIRVWriter.cpp [hunk 3]
```diff
@@ -3778,27 +3801,18 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
...
-              DecSubj, AnnotationString.c_str()));
-        else
-          // Apply the LSU parameter decoration to the pointer result of an
-          // instruction. Note it's the address to the accessed memory that's
-          // loaded from the original pointer variable, and not the value
-          // accessed by the latter.
-          addAnnotationDecorations(DecSubj, Decorations.MemoryAccessesVec);
-        II->replaceAllUsesWith(II->getOperand(0));
-      }
+      // Memory accesses to a standalone pointer variable
```
Comment (asudarsa) [inline lib/SPIRV/SPIRVWriter.cpp:3804]: Can you please specify if the scope of this change goes beyond addressing the first member of the struct variable? This change also seems unrelated. Thanks

File: lib/SPIRV/SPIRVWriter.cpp (hunk 2 above)
Comment (dwoodwor-intel) [reply lib/SPIRV/SPIRVWriter.cpp:3748]: This was something I discovered when running the tests: there are a few cases where this FPGARegIntel annotation is used on struct values, and my change was incorrectly trying to translate those as OpMemberDecorate instead. It seems like in cases where we have an FPGARegIntel annotation we want to translate it to OpFPGARegINTEL regardless of whether we find a struct type, so I thought the cleanest way of doing that would be moving this check up here.

@vmaksimo, from the git log it looks like you originally added this code; is putting it here correct?

File: lib/SPIRV/SPIRVWriter.cpp (hunk 3 above)
Comment (dwoodwor-intel) [reply lib/SPIRV/SPIRVWriter.cpp:3804]: I've just removed the OpFPGARegINTEL code from here; the rest of the changes in this part are just formatting

File: lib/SPIRV/SPIRVWriter.cpp (hunk 1 above)
Comment (dwoodwor-intel) [reply lib/SPIRV/SPIRVWriter.cpp]: Done

File: lib/SPIRV/SPIRVWriter.cpp (hunk 3 above)
Comment (asudarsa) [reply lib/SPIRV/SPIRVWriter.cpp:3804]: Looks good. Thanks

**Review by MrSidims — APPROVED** [review]
CI issues don't relate with the patch
Checking them

**Review by asudarsa — APPROVED** [review]
LGTM. Please validate SPIR-V generated using spirv-val. Thanks

**Comment by dwoodwor-intel** [comment]
> LGTM. Please validate SPIR-V generated using spirv-val. Thanks

I ran spirv-val on this test and did find an error:

```
$ spirv-val annotate_attribute.spv
error: line 236: Variables can not have a function[7] storage class outside of a function
  %_str = OpVariable %_ptr_Function__arr_uchar_ulong_3 Function %9
```

I also tried without this change on the main branch and I see the same error there too. It seems to be a preexisting issue unrelated to this change.

File: lib/SPIRV/SPIRVWriter.cpp (hunk 2 above)
Comment (vmaksimo) [reply lib/SPIRV/SPIRVWriter.cpp:3748]: @dwoodwor-intel Oh, that has been some long time ago, and I missed a big part of changes related to this feature transformation :) 

But I do think that now it's a correct place - it looks logical and has not failed any tests.

**Comment by MrSidims** [comment]
llvm-bolt is a known issue reported in https://github.com/llvm/llvm-project/issues/58317 and several other places. Merging.
